### ISSUE TITLE
Omit is breaking `in` gaurd

### ISSUE THREAD

**Issue by stavalfi** [microsoft/TypeScript#32030]
**TypeScript Version:**  3.5.2

**Search Terms: Omit, guard**

**Code**
```ts
type t1 = { a: boolean } & ({ x: number } | { y: string } | {})
const z1: t1 = Date.now() === 123453 ? { a: true, x: 1 } : { a: true }
const z2: Omit<t1, 'a'> = z1
console.log('x' in z2 && z2.x)
```

**Expected behavior:**
no errors
**Actual behavior:**
> Error:(235, 29) TS2339: Property 'x' does not exist on type 'never'.

**Playground Link:** 
[Link to Playground](https://www.typescriptlang.org/play/#src=%2F%2F%20Omit%20copied%20from%20typescript%20lib.es5.d.ts%20file%0D%0Atype%20Omit%3CT%2C%20K%20extends%20keyof%20any%3E%20%3D%20Pick%3CT%2C%20Exclude%3Ckeyof%20T%2C%20K%3E%3E%0D%0A%0D%0Atype%20t1%20%3D%20%7B%20a%3A%20boolean%20%7D%20%26%20(%7B%20x%3A%20number%20%7D%20%7C%20%7B%20y%3A%20string%20%7D%20%7C%20%7B%7D)%0D%0Aconst%20z1%3A%20t1%20%3D%20Date.now()%20%3D%3D%3D%20123453%20%3F%20%7B%20a%3A%20true%2C%20x%3A%201%20%7D%20%3A%20%7B%20a%3A%20true%20%7D%0D%0Aconst%20z2%3A%20Omit%3Ct1%2C%20'a'%3E%20%3D%20z1%0D%0Aconsole.log('x'%20in%20z2%20%26%26%20z2.x))

-----

If I specify the actual type instead of using Omit, everything is working with no errors:

```ts
type t1 = { a: boolean } & ({ x: number } | { y: string } | {})
const z1: t1 = Date.now() === 123453 ? { a: true, x: 1 } : { a: true }
const z2: { x: number } | { y: string } | {} = z1
console.log('x' in z2 && z2.x)
```

**Comment by jack-williams** [comment]
For this to work you need to implement you `Omit` using a distributive `Pick`. See #28339 for extensive discussion.

TLDR:
```ts
// Omit copied from typescript lib.es5.d.ts file
type UnionKeys<T> = T extends any ? keyof T : never
type DistributivePick<T, K extends UnionKeys<T>> = T extends any ? Pick<T, Extract<keyof T, K>> : never;
type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>
type DistributiveOmit<T, K extends UnionKeys<T>> = T extends any ? Omit<T, Extract<keyof T, K>> : never;

type t1 = { a: boolean } & ({ x: number } | { y: string } | {})
const z1: t1 = Date.now() === 123453 ? { a: true, x: 1 } : { a: true }
const z2: DistributiveOmit<t1, 'a'> = z1
console.log('x' in z2 && z2.x)
```

**Comment by stavalfi** [comment]
Okay. You gave me a lot to learn before I can respond. 

If this is working as Intended, close this issue.
